pkg/rc/replication_controller.go

```diff
@@ -654,23 +654,66 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, rcFields f
+//      replicas than current (described more in canNodeTransfer())
 // If the conditions are not met, the function will be called again on the next
 // call of meetDesires() should a node still be ineligible. It returns true
 // when a node transfer occurs
-func (rc *replicationController) attemptNodeTransfer(rcFields fields.RC, current types.PodLocations, ineligible types.NodeName, allocAttempts int) (bool, error) {
+func (rc *replicationController) attemptNodeTransfer(rcFields fields.RC, current types.PodLocations, ineligibles []types.NodeName, allocAttempts int) (bool, error) {
+	ok, unlocker, err := rc.canNodeTransfer(rcFields, current, ineligibles)
+	if err != nil {
+		return false, err
+	} else if !ok {
+		return false, nil
+	}
+	if unlocker != nil {
+		defer unlocker.Unlock()
+	}
+
+	ineligible := ineligibles[0]
+	err = rc.swapNodes(rcFields, current, ineligible, allocAttempts)
+	if err != nil {
+		rc.logger.WithError(err).Errorln("could not swap nodes")
+		alertErr := rc.alerter.Alert(rc.alertInfo(rcFields, err.Error()), alerting.LowUrgency)
+		if alertErr != nil {
+			rc.logger.WithError(alertErr).Errorln("unable to send alert")
+		}
+		return false, err
+	}
+
+	rc.logger.Infof("transferred off %s", ineligible)
+	return true, nil
+}
+
+func (rc *replicationController) canNodeTransfer(rcFields fields.RC, current types.PodLocations, ineligibles []types.NodeName) (bool, consul.Unlocker, error) {
+	currentSet := types.NewNodeSet(current.Nodes()...)
+	ineligibleSet := types.NewNodeSet(ineligibles...)
+	if rcFields.Disabled &&
+		currentSet.Equal(ineligibleSet) &&
+		rcFields.ReplicasDesired < len(current) {
+		// If the RC is disabled, all of its nodes are inelgible, and its
+		// desired replica count is less than its current replica count,
+		// then it is likely the old RC in an RU. In that case, the new RC
+		// will not be able to schedule on the ineligible nodes, so we should
+		// perform a node transfer. The RU would not have decremented this RC's
+		// replicas desired unless its min health was met, so we can safely
+		// perform the swap.
+		return true, nil, nil
+	}
+	ineligible := ineligibles[0]
 	ok, err := rc.isTransferMinHealthMet(rcFields, current, ineligible)
 	if err != nil {
 		rc.logger.WithError(err).Errorln("skipping node transfer; error checking health")
-		return false, err
+		return false, nil, err
 	} else if !ok {
 		rc.logger.Infoln("skipping node transfer; node transfer health requirement not met")
-		return false, nil
+		return false, nil, nil
 	}
 
 	_, session, err := consul.SessionContext(context.Background(), rc.consulClient, fmt.Sprintf("rc-node-transfer-%s", rc.rcID))
 	if err != nil {
 		rc.logger.WithError(err).Errorln("could not create a session to acquire rc lock")
-		return false, err
+		return false, nil, err
 	}
 	// Node transfers do not mutate the RC's state, but we attempt to acquire
 	// the lock to ensure that we do not perform a transfer during a rolling
```

Comment:
We probably need to polish the annotations here to diminish the confusion. I suppose you mean something like, to ensure that we do not perform a rolling update during node transfer.